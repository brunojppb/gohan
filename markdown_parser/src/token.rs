use std::fmt::{Debug, Display};

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Token {
    pub position: usize,
    pub kind: TokenType,
    pub line: usize,
}

impl Token {
    pub fn new(position: usize, kind: TokenType, line: usize) -> Self {
        Self {
            position,
            kind,
            line,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum TokenType {
    Hash,
    Star,
    Bang,
    Underscore,
    Newline,
    Tab,
    Space,
    Dot,
    Dash,
    Backslash,
    LeftParen,
    RightParen,
    LeftSquareBraket,
    RightSquareBraket,
    Number(usize),
    Text(String),
    EndOfFile,
}

impl TokenType {
    /// Lexeme representation. Mostly for debugging
    pub fn lexime(&self) -> String {
        match self {
            Self::Hash => "#".to_string(),
            Self::Star => "*".to_string(),
            Self::Bang => "!".to_string(),
            Self::Dot => ".".to_string(),
            Self::Dash => "-".to_string(),
            Self::Backslash => "\\".to_string(),
            Self::LeftParen => "(".to_string(),
            Self::RightParen => ")".to_string(),
            Self::LeftSquareBraket => "[".to_string(),
            Self::RightSquareBraket => "]".to_string(),
            Self::Tab => "\\t".to_string(),
            Self::Space => " ".to_string(),
            Self::Newline => "\\n".to_string(),
            Self::Underscore => "_".to_string(),
            Self::Number(number) => number.to_string(),
            Self::Text(text) => text.to_string(),
            Self::EndOfFile => "EOF".to_string(),
        }
    }
}

impl Display for TokenType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "lexeme={}", self.lexime())
    }
}
