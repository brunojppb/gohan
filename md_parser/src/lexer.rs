use crate::token::{Span, Token};

const SYMBOLS: &str = "#*!_[]().- \n\t\\";

/// Tokenizes Markdown input
pub struct Lexer<'a> {
    source: &'a str,
    tokens: Vec<(Token<'a>, Span)>,
    start_byte_offset: usize,
    current_byte_offset: usize,
    col: usize,
    line: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            source: input,
            tokens: Vec::new(),
            start_byte_offset: 0,
            current_byte_offset: 0,
            line: 1,
            col: 1,
        }
    }

    pub fn scan(&mut self) -> &Vec<(Token<'a>, Span)> {
        while !self.is_at_end() {
            self.start_byte_offset = self.current_byte_offset;
            self.scan_token();
        }

        self.add_token(Token::EndOfFile);
        &self.tokens
    }

    fn scan_token(&mut self) {
        let Some(c) = self.advance() else {
            panic!(
                "Could not scan the next token. line={} byte_offset={}",
                &self.line, self.current_byte_offset
            );
        };

        match c {
            b'#' => self.add_token(Token::Hash),
            b'*' => self.add_token(Token::Star),
            b'!' => self.add_token(Token::Bang),
            b' ' => self.add_token(Token::Space),
            b'_' => self.add_token(Token::Underscore),
            b'-' => self.add_token(Token::Dash),
            b'.' => self.add_token(Token::Dot),
            b'(' => self.add_token(Token::LeftParen),
            b')' => self.add_token(Token::RightParen),
            b'[' => self.add_token(Token::LeftSquareBracket),
            b']' => self.add_token(Token::RightSquareBracket),
            b'\\' => self.add_token(Token::Backslash),
            b'\t' => self.add_token(Token::Tab),
            b'\n' => self.add_token(Token::Newline),
            c if c.is_ascii_digit() => self.add_token(Token::Digit(
                &self.source[self.current_byte_offset - 1..self.current_byte_offset],
            )),
            _ => self.handle_string(),
        }
    }

    fn is_token(&self, c: Option<u8>) -> bool {
        match c {
            Some(c) => {
                if c.is_ascii() {
                    c.is_ascii_digit() || SYMBOLS.contains(c as char)
                } else {
                    false
                }
            }
            None => false,
        }
    }

    fn handle_string(&mut self) {
        let start_offset = self.current_byte_offset - 1;
        let mut end_byte_offset = start_offset;
        while !self.is_at_end() && !self.is_token(self.peek()) {
            self.advance();
            end_byte_offset += 1;
        }

        let value = &self.source[start_offset..end_byte_offset + 1];

        self.add_token(Token::Text(value));
    }

    fn is_at_end(&self) -> bool {
        self.current_byte_offset >= self.source.bytes().len()
    }

    fn add_token(&mut self, token: Token<'a>) {
        // let token = Token::new(self.start, token_type, self.line);
        let span = Span {
            line: self.line,
            col: self.col,
        };
        self.tokens.push((token, span));
    }

    /// Look-up the next character, but do not consume it
    fn peek(&self) -> Option<u8> {
        if self.is_at_end() {
            return None;
        }
        self.source
            .as_bytes()
            .get(self.current_byte_offset)
            .copied()
    }

    /// Consume the next byte and advance the needle
    /// to point to a potential next character.
    /// byte continution of multi-byte characters
    /// should be handled by the caller.
    fn advance(&mut self) -> Option<u8> {
        if let Some(c) = self
            .source
            .as_bytes()
            .get(self.current_byte_offset)
            .copied()
        {
            if c == b'\n' {
                self.line += 1;
                self.col = 1;
            } else {
                self.col += 1;
            }

            self.current_byte_offset += 1;
            return Some(c);
        }

        self.current_byte_offset += 1;
        None
    }
}

#[cfg(test)]
mod tests {
    use std::fs;

    use super::*;

    #[test]
    fn tokenize_markdown() {
        insta::glob!("snapshot_inputs/*.md", |path| {
            let markdown = fs::read_to_string(path).unwrap();
            let mut lexer = Lexer::new(&markdown);
            insta::assert_json_snapshot!(lexer.scan());
        });
    }

    #[test]
    fn accept_multi_byte_chars() {
        let markdown = r"
        ## This is a title
        
        This should include emojis and **bold text**.
        🤡😜🎉 text 🐙👪.
        ";
        let mut lexer = Lexer::new(markdown);
        let result = lexer.scan();
        assert_eq!(result.len(), 80);
    }
}
