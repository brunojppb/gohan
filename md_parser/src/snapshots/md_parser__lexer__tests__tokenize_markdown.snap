---
source: md_parser/src/lexer.rs
expression: lexer.scan()
input_file: md_parser/src/snapshot_inputs/lexer_input.md
---
[
  [
    "Hash",
    {
      "line": 1,
      "col": 1
    }
  ],
  [
    "Space",
    {
      "line": 1,
      "col": 2
    }
  ],
  [
    {
      "Text": "Hi"
    },
    {
      "line": 1,
      "col": 4
    }
  ],
  [
    "Space",
    {
      "line": 1,
      "col": 5
    }
  ],
  [
    {
      "Text": "there"
    },
    {
      "line": 1,
      "col": 10
    }
  ],
  [
    "Newline",
    {
      "line": 2,
      "col": 0
    }
  ],
  [
    "Newline",
    {
      "line": 3,
      "col": 0
    }
  ],
  [
    {
      "Text": "This"
    },
    {
      "line": 3,
      "col": 4
    }
  ],
  [
    "Space",
    {
      "line": 3,
      "col": 5
    }
  ],
  [
    {
      "Text": "should"
    },
    {
      "line": 3,
      "col": 11
    }
  ],
  [
    "Space",
    {
      "line": 3,
      "col": 12
    }
  ],
  [
    {
      "Text": "be"
    },
    {
      "line": 3,
      "col": 14
    }
  ],
  [
    "Space",
    {
      "line": 3,
      "col": 15
    }
  ],
  [
    {
      "Text": "text"
    },
    {
      "line": 3,
      "col": 19
    }
  ],
  [
    "Space",
    {
      "line": 3,
      "col": 20
    }
  ],
  [
    "Underscore",
    {
      "line": 3,
      "col": 21
    }
  ],
  [
    {
      "Text": "wrapped"
    },
    {
      "line": 3,
      "col": 28
    }
  ],
  [
    "Space",
    {
      "line": 3,
      "col": 29
    }
  ],
  [
    {
      "Text": "in"
    },
    {
      "line": 3,
      "col": 31
    }
  ],
  [
    "Space",
    {
      "line": 3,
      "col": 32
    }
  ],
  [
    {
      "Text": "underscore"
    },
    {
      "line": 3,
      "col": 42
    }
  ],
  [
    "Underscore",
    {
      "line": 3,
      "col": 43
    }
  ],
  [
    "Space",
    {
      "line": 3,
      "col": 44
    }
  ],
  [
    {
      "Text": "and"
    },
    {
      "line": 3,
      "col": 47
    }
  ],
  [
    "Space",
    {
      "line": 3,
      "col": 48
    }
  ],
  [
    "Backslash",
    {
      "line": 3,
      "col": 49
    }
  ],
  [
    "Backslash",
    {
      "line": 3,
      "col": 50
    }
  ],
  [
    {
      "Text": "escaped"
    },
    {
      "line": 3,
      "col": 57
    }
  ],
  [
    "Newline",
    {
      "line": 4,
      "col": 0
    }
  ],
  [
    "Star",
    {
      "line": 4,
      "col": 1
    }
  ],
  [
    "Star",
    {
      "line": 4,
      "col": 2
    }
  ],
  [
    {
      "Text": "bold"
    },
    {
      "line": 4,
      "col": 6
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 7
    }
  ],
  [
    {
      "Text": "text"
    },
    {
      "line": 4,
      "col": 11
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 12
    }
  ],
  [
    {
      "Text": "here"
    },
    {
      "line": 4,
      "col": 16
    }
  ],
  [
    "Star",
    {
      "line": 4,
      "col": 17
    }
  ],
  [
    "Star",
    {
      "line": 4,
      "col": 18
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 19
    }
  ],
  [
    {
      "Text": "with"
    },
    {
      "line": 4,
      "col": 23
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 24
    }
  ],
  [
    {
      "Text": "a"
    },
    {
      "line": 4,
      "col": 25
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 26
    }
  ],
  [
    {
      "Text": "nice"
    },
    {
      "line": 4,
      "col": 30
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 31
    }
  ],
  [
    "LeftSquareBracket",
    {
      "line": 4,
      "col": 32
    }
  ],
  [
    {
      "Text": "link"
    },
    {
      "line": 4,
      "col": 36
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 37
    }
  ],
  [
    {
      "Text": "here"
    },
    {
      "line": 4,
      "col": 41
    }
  ],
  [
    "RightSquareBracket",
    {
      "line": 4,
      "col": 42
    }
  ],
  [
    "LeftParen",
    {
      "line": 4,
      "col": 43
    }
  ],
  [
    {
      "Text": "https://bpaulino"
    },
    {
      "line": 4,
      "col": 59
    }
  ],
  [
    "Dot",
    {
      "line": 4,
      "col": 60
    }
  ],
  [
    {
      "Text": "com"
    },
    {
      "line": 4,
      "col": 63
    }
  ],
  [
    "RightParen",
    {
      "line": 4,
      "col": 64
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 65
    }
  ],
  [
    {
      "Text": "and"
    },
    {
      "line": 4,
      "col": 68
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 69
    }
  ],
  [
    {
      "Text": "text"
    },
    {
      "line": 4,
      "col": 73
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 74
    }
  ],
  [
    {
      "Text": "after"
    },
    {
      "line": 4,
      "col": 79
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 80
    }
  ],
  [
    {
      "Text": "the"
    },
    {
      "line": 4,
      "col": 83
    }
  ],
  [
    "Space",
    {
      "line": 4,
      "col": 84
    }
  ],
  [
    {
      "Text": "link"
    },
    {
      "line": 4,
      "col": 88
    }
  ],
  [
    "Newline",
    {
      "line": 5,
      "col": 0
    }
  ],
  [
    "EndOfFile",
    {
      "line": 5,
      "col": 0
    }
  ]
]
